PHAsset 프레임워크 적용하기 #4

GangWoon opened the pull request:
collectionView cell에 이미지를 추가하기 위해서 UICollectionViewCell을 상속받는 커스텀 클래스를 만들어서 imageView와 imageView의 이미지를 변경하는 메소드를 만들었습니다.
Photos 프레임워크를 사용해서 collecitonView cell에 이미지를 표시했습니다.
PHAsset을 생성할때 PHFetchOptions을 외부에서 받아오게 만들고 싶었지만, Photo를 다른 파일에서도 내포해야되서 한 클래스에서 처리하도록 결정했습니다.

1Consumption committed:
[기능 추가] PhotoCell 클래스 생성

각 cell의 imageView에 접근하기위해 IBOulet을 연결받는 클래스 생성

GangWoon committed:
[기능 수정] 불필요한 사진 삭제

더이상 collectionView를 확인할 사진이 필요없어서 삭제했습니다.

GangWoon committed:
[기능 추가] 사용자 앨범 접근

사용자 앨범에 접근할 수 있도록 Info.plist 파일을 변경했습니다.
추가적인 코맨트는 남기지 않았습니다.

GangWoon committed:
[기능 수정] 컨밴션 정리

GangWoon committed:
[기능 수정] CollectionViewDataSource 초기화 수정

더이상 CollectionViewDataSource을 초기화할때 size 속성을 초기화 할 필요가 없어져서 수정했습니다.

GangWoon committed:
[기능 추가] CollectionViewCell에 사진 표시

Photos를 처리하는 객체를 만들건지 고민했었습니다.
객체를 만들어서 처리한다면 cell를 photos객체로 넘겨야했고, collectionViewCell을 처리하기 위해서 UIKit을 내포해야됩니다.
더 확신이 든 이유는 requestImage(for: , targetSize: , contentMode: , options: , resultHandler: )를 사용해서 이미지를 설정해야 되는데 cell을 받아오던가, closure로 넘겨서 처리를 하는 부분이 너무 부자연스러웠습니다.

GangWoon committed:
[파일 변경 커밋]

GangWoon committed:
[기능 수정] 속성들의 타입 변경

!을 잘못 사용하고 있어서 변경했습니다.
원래 의도는 init()에서 처리해주는 작업들을 메소드로 만들어서 처리하고 싶었지만 나누다보니 이상하다는 느낌을 많이 받았습니다.
다시 코드를 고치는 작업하는중에 놓친 부분을 발견해서 변경했습니다.

godrm commented:
커밋 로그들을 보면 모두 기여하지 않은 것처럼 한 사람만 주로 보이는 이유는 무엇인가요?

godrm commented on PhotosApp/PhotosApp/Info.plist:
```diff
@@ -28,6 +28,8 @@
 		<string>armv7</string>
 	</array>
+	<key>NSPhotoLibraryUsageDescription</key>
+	<string></string>
```
사용하는 문구를 꼭 명시하세요. 이게 사용자에게 그대로 보입니다.

godrm commented on PhotosApp/PhotosApp/Model/CollectionViewDataSource.swift:
```diff
@@ -7,27 +7,43 @@
+        photoOptions.sortDescriptors = [
+            NSSortDescriptor(key: "creationDate", ascending: true)
+        ]
+        fetchResult = PHAsset.fetchAssets(with: photoOptions)
```
이렇게 정렬해서 모두 fetch하는게 앱 실행하는 시점에서 사진이 많으면 오래 걸리지 않나요? 
바로 화면이 뜨는지 직접 (사진이 많은) 디바이스에서 실행해서 확인하세요.

GangWoon replied:
아이폰 X에서 실행한 결과 사진이 약 1400장 정도 있었는데 바로 실행이 되는 모습을 확인했습니다.
그런데 엄청나게 많은 데이터가 있다면 느려질 수 있다고 생각했습니다.
그래서 저희가 생각한 방법은 PHFetchOptions.fetchLimit 속성을 변경하고 특정 이벤트(스크롤의 위치가 마지막인 경우와 더 받아올 데이터가 존재한 경우)가 발생하면  더 받아오도록 구현해야 될 거 같습니다. 하지만 구현은 못 했습니다. 어떻게 생각하시나요?

godrm replied:
PHFetchOptions.fetchLimit 를 하고 prefetch 정도를 같이 하면 적당합니다. 
실제로 해보는게 어렵지 않으면 해보세요.

godrm replied:
이런게 전형적으로 데이터가 많을 수록 느려지는 부분이니까 전체 데이터를 다 가져오는 건 항상 주의해야 합니다.

GangWoon commented:
울라프의 컨디션이 안좋아서 혼자서 코딩을 했었습니다.
페어 프로그래밍을하기 위해서 더 노력하겠습니다.

godrm commented on PhotosApp/PhotosApp/Model/CollectionViewDataSource.swift:
```diff
@@ -7,27 +7,43 @@
     public func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
-        let cell = collectionView.dequeueReusableCell(
+        guard let cell = collectionView.dequeueReusableCell(
             withReuseIdentifier: UICollectionViewCell.photosCellIdentifier,
```
Custom 셀 이름이 PhotoCell 인데 Identifier는 `UICollectionViewCell.photosCellIdentifier` 네요? PhotoCell과 관련이 있는게 아닐까요?

GangWoon replied:
리뷰를 신청하기 전에 페어와 충분히 코드를 검토한 후에 리뷰를 신청하겠습니다.
너무 마음만 서둘러서 생긴 실수였습니다.

godrm replied:
실수를 통해 배우면 되죠. 스스로 인지할 수 있도록 코딩할 때 규칙을 정해보세요.

godrm commented on PhotosApp/PhotosApp/Model/CollectionViewDataSource.swift:
```diff
@@ -7,27 +7,43 @@
+            contentMode: .aspectFill,
+            options: .none
+        ) { image, _ in
+            cell.apply(with: image)
```
여기 클로저는 동기 방식인가 비동기 방식인가요? 
셀이 재사용되는 데 이미지를 이렇게 적용해놓으면 스크롤 될때 문제되지는 않나요?

1Consumption replied:
비동기 방식입니다.

requestImage(for:, targetSize:, contentMode:, options:, resultHandler:) 메소드에서 resultHandler가 최초 저화질의 이미지를 셀에 넣어주고, 화면에 보이는 모든 셀에 이미지가 들어간 후 다시 비동기적으로 고화질의 이미지를 넣어주는데, 지나치게 빠르게 스크롤할 경우 저화질의 이미지를 넣고 고화질의 이미지를 넣으려고 하는 순간, 셀이 바뀌면서 기대했던 이미지가 아닌, 이전 인덱스패스에 해당하는 고화질 이미지가 현재 셀에 들어가기 때문에 이상한 이미지가 들어갑니다.(저희들 생각입니다)
이를 보완하기 위해 현재 인덱스패스와 넣고자하는 이미지의 인덱스값과 같은지를 비교하고, 같으면 이미지를 넣어주도록 했습니다. 
그 결과 스크롤을 빠르게 해도 이미지가 꼬이지않고 잘 들어가는것을 확인했습니다.

godrm commented on PhotosApp/PhotosApp/View/Base.lproj/Main.storyboard:
```diff
@@ -26,15 +26,15 @@
-                                                <imageView clipsSubviews="YES" userInteractionEnabled="NO" contentMode="scaleAspectFit" horizontalHuggingPriority="251" verticalHuggingPriority="251" image="마동석" translatesAutoresizingMaskIntoConstraints="NO" id="U6m-YG-leP">
-                                                    <rect key="frame" x="0.0" y="0.0" width="100" height="100.00000000000003"/>
+                                                <imageView clipsSubviews="YES" userInteractionEnabled="NO" contentMode="scaleAspectFit" horizontalHuggingPriority="251" verticalHuggingPriority="251" translatesAutoresizingMaskIntoConstraints="NO" id="U6m-YG-leP">
+                                                    <rect key="frame" x="0.0" y="0.0" width="100" height="100"/>
```
이렇게 스토리보드에 이미지뷰 크기를 고정하면 Cell 크기를 늘리거나 작게 바꾸면 같이 줄어드나요?

1Consumption replied:
넵 오토레이아웃 적용해서 가변적으로 잘 대응합니다

godrm reviewed (changes requested):
다음 피드백 내용을 확인해보고 resolve 한 다음에 머지하세요.

godrm commented:
> 울라프의 컨디션이 안좋아서 혼자서 코딩을 했었습니다.
> 페어 프로그래밍을하기 위해서 더 노력하겠습니다.

그랬다면 구현 내용을 모를테니, 올라프에게도 리뷰를 받고 머지하세요.

1Consumption commented:
더 이상 generateRandomColor를 사용하지 않는데 왜 아직 있을까요?

GangWoon commented:
놓친 부분이었습니다. 삭제하도록 하겠습니다.

godrm reviewed (approved):
서로 확인했으면 머지하고 진행하시죠.

GangWoon committed:
[기능 수정] 접근 권한 코멘트 변경

사용자가 처음 앱을 켰을 때 앨범 접근 권한을 명시적으로 확인할 수 있도록 변경했습니다.

GangWoon committed:
[기능 수정] 사용하지 않는 속성 제거

스토리보드에서 크기와 관련된 작업을 처리했기 때문에 제거했습니다.

GangWoon committed:
[기능 수정] 사용하지 않는 속성 제거

셀을 표시할 때 더 이상 백그라운드 컬러를 변경할 필요가 없기 때문에 제거했습니다.

GangWoon committed:
[기능 수정] Identifier의 위치 변경

cell 타입이 UICollectionViewCell이 아닌 PhotoCell이기 때문에 해당 extension을 삭제했습니다.

GangWoon committed:
[기능 추가] PhotoCell Identifer 추가

GangWoon committed:
[기능 추가] Photo 프레임워크를 처리하는 클래스 생성

datasource에서는 메타데이터가 필요하다고 생각했고, 그걸 처리해주는 객체를 만드는게 적합하다 생각했습니다.

GangWoon committed:
[기능 수정] PhotoManager 객체 생성

역할을 분리했기 때문에 그에 적합하게 코드를 변경했습니다.

GangWoon committed:
[파일 생성 및 삭제 커밋]

GangWoon committed:
[기능 수정] 이미지 컨텐트 모드 변경

이미지 뷰의 이미지를 가득 채우기 위해 변경했습니다.

1Consumption committed:
[기능 수정] 사진앱 접근 권한 설정

사진 앱 권한 설정 후 asset이 받아와지지 않던 오류를 수정했습니다.
이는 notDetermined 상태에서 authorized상태가 되어도 접근 권한이 없었을 때 불러왔던 asset이 유지되기 때문에,
reloadAsset 메소드를 dataSource에 만들어서 권한이 있는 상태에서 asset을 다시 요청했습니다.
또한 view를 바꾸는 과정이기 때문에 main스레드에서 aync하게 처리했습니다.

GangWoon commented:
PHFetchOptions.fetchLimit 와 prefetch를 사용해서 구현하고 넘어가세요

1Consumption committed:
[기능 수정]  이미지 삽입 방법 개선

현재 인덱스패스와 넣고자하는 이미지의 인덱스값과 같은지를 비교하고, 같으면 이미지를 넣어주도록 수정
